### PR TITLE
ISPN-9841 Non-blocking Hot Rod authentication

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-9841

### CODE REVIEW

**Comment by tristantarrant** [comment]
3 related failures

**Comment by tristantarrant** [comment]
Fixed the failures by reordering how we handle "complete" authentication. Before sending the final server response, we obtain the subject from the auth backend.

**Comment by rvansa** [comment]
There are couple of places where the auth process throws an exception. In the original code this was handled in the server (writing the error response) but I am afraid here it just gets logged and the client does not get any response until it times out. Could you make sure to write some response on any code path?

**Comment by tristantarrant** [comment]
@rvansa are you sure ? I'm just wrapping the exceptions into a SecurityException which will then be handled by HotRodDecoder.decode()'s catch

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java [hunk 1]
```diff
@@ -86,74 +75,105 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
...
+            ssf = server.getSaslServerFactory(mech);
+         }
+         if (authenticationConfig.serverSubject() != null) {
+            try {
                saslServer = Subject.doAs(authenticationConfig.serverSubject(), (PrivilegedExceptionAction<SaslServer>) () ->
                      ssf.createSaslServer(mech, "hotrod", authenticationConfig.serverName(),
                            authenticationConfig.mechProperties(), callbackHandler));
-            } else {
+            } catch (PrivilegedActionException e) {
+               throw new SecurityException(writeException(header, e.getCause()));
```
Comment (rvansa) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java]: Why are you throwing that here? Can't you just log the PAE, writeException and `return`?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java [hunk 2]
```diff
@@ -86,74 +75,105 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
...
+            }
+         } else {
+            try {
                saslServer = ssf.createSaslServer(mech, "hotrod", authenticationConfig.serverName(),
                      authenticationConfig.mechProperties(), callbackHandler);
+            } catch (SaslException e) {
+               throw new SecurityException(writeException(header, e));
```
Comment (rvansa) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java]: In general I find calling something with side-effect to exception ctor a bit cryptic (couldn't spot where it is written for a while), though YMMV.

**Comment by tristantarrant** [comment]
Yeah, that was really silly of me. I've changed  that @rvansa

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java [hunk 3]
```diff
@@ -86,74 +75,109 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
...
-               SaslQopHandler qopHandler = new SaslQopHandler(saslServer);
-               channel.pipeline().addBefore("decoder", "saslQop", qopHandler);
-            } else {
-               saslServer.dispose();
-               callbackHandler = null;
-               saslServer = null;
-            }
+      writeResponse(header, header.encoder().authResponse(header, server, channel.alloc(), serverChallenge));
+
+      if (saslServer.isComplete()) {
```
Comment (rvansa) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java]: Isn't here a race between responding with the authResponse and adding the QOP handler below?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java [hunk 4]
```diff
@@ -86,74 +75,109 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
...
-            if ("EXTERNAL".equals(mech)) {
-               SslHandler sslHandler = channel.pipeline().get(SslHandler.class);
-               try {
-                  if (sslHandler != null)
-                     ssf = new ExternalSaslServerFactory(sslHandler.engine().getSession().getPeerPrincipal());
-                  else
-                     throw log.externalMechNotAllowedWithoutSSLClientCert();
-               } catch (SSLPeerUnverifiedException e) {
-                  throw log.externalMechNotAllowedWithoutSSLClientCert();
+         executor.execute(() -> authInternal(header, mech, response));
```
Comment (rvansa) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java]: While you have handled all the declared exceptions, I would still feel a bit more warm&fuzzy if any unexpected exceptions (NPEs...) would be handled with a response...

**Comment by tristantarrant** [comment]
@rvansa I believe I've addressed everything

**Comment by tristantarrant** [comment]
Kicked-off a new CI build as the previous one crashed in core (unrelated to this PR)

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java (hunk 3 above)
Comment (danberindei) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java]: @rvansa @tristantarrant isn't this still valid?

**Comment by tristantarrant** [comment]
A green build !
:bowtie:

**Pull request merged**